File: implement-shell-tools/ls/ls.mjs
import { program } from "commander";
import { promises as fs} from "node:fs"

const TERMINAL_WIDTH = 120;
const NAME_PADDING = 8;

program
  .name("list")
  .description("list directory contents")
  .option("-a", "Include directory entries whose names begin with a dot (‘.’).")
  .option("-1", "(The numeric digit “one”.) Force output to be one entry per line.")
  .argument("[path...]", "path to entries to list");

program.parse()

const argv = program.args;
const options = program.opts();

start();

async function start() {
  const output = [];
  const files = [];
  const directories = [];
  
  await checkInput(output, files, directories);
  if (files.length) {
    output.push(formatFilesOutput(files));
  }
  await formatDirectoriesOutput(output, directories, files.length);

  for(const string of output) {
    if (string.startsWith("ls.mjs")) {
      console.error(string);
    } else {
      console.log(string);
    }
  }
}

async function checkInput(output, files, directories) {
  if (!argv.length) {
    directories.push(".");
    return;
  }
  for (const path of argv) {
    try {
      const stat = await fs.stat(path);
      if (stat.isFile()) {
        files.push(path);
      }
      if (stat.isDirectory()) {
        directories.push(path);
      }
    } catch (error) {
      output.push(`ls.mjs:  ${path}: No such file ore directory`);
    }
  }
  files.sort();
  directories.sort();
}

function formatFilesOutput(files) {
  if (options[1]) {
    return files.join("\n");
  }
  let maxLength = 0;
  for (const name of files) {
    if (maxLength < name.length) {
      maxLength = name.length;
    }
  }

  let padLength = 0;
  while (padLength < maxLength) {
    padLength += NAME_PADDING;
  }

  let output = "";
  let lineLength = 0;
  for (const name of files) {
    output += name.padEnd(padLength, " ");
    lineLength += padLength;
    if (TERMINAL_WIDTH - lineLength < padLength) {
      output += "\n";
      lineLength = 0;
    }
  }
  return output;
}

async function formatDirectoriesOutput(output, directories, isFilesExist) {
  const isSingleDirectory = directories.length === 1;
  for(let i = 0; i < directories.length; i++) {
    let files = await fs.readdir(directories[i]);
    if (options.a) {
      files.push(".", "..");
    } else {
      files = files.filter(file => !file.startsWith("."));
    }
    files.sort();
    let directoryOutput = formatFilesOutput(files);
    if (isFilesExist || !isSingleDirectory) {
      directoryOutput = `${directories[i]}:\n${directoryOutput}`;
    }
    if (isFilesExist || !isSingleDirectory && i) {
      directoryOutput = `\n${directoryOutput}`;
    }
    output.push(directoryOutput);
  }
}

File: implement-shell-tools/ls/ls.py
import os
import sys
import argparse

TERMINAL_LENGTH = 120

parser = argparse.ArgumentParser(
  prog = "list-directory-content",
  description = "For each operand that names a file of a type other than directory, ls displays its name as well as any requested, associated" \
    "information.  For each operand that names a file of type directory, ls displays the names of files contained within that directory, " \
    "as well as any requested, associated information."
)

parser.add_argument("path", help = "path to entries to list", nargs = "+")
args = parser.parse_args()


def start():
  output = []
  files = []
  directories = []

  checkInput(output, files, directories)

  if (len(files)):
    output.append(formatFilesOutput(files))

  printOutput(output)

def checkInput(output, files, directories):
  for path in args.path:
    if (os.path.isdir(path)):
      directories.append(path)
    elif (os.path.isfile(path)):
      files.append(path)
    else:
      output.append(f"ls.py:  {path}: No such file ore directory")

  files.sort()
  directories.sort()

def formatFilesOutput(files):
  maxLength = 0
  for name in files:
    if (maxLength < len(name)):
      maxLength = len(name)
  
  padLength = 0
  while (padLength < maxLength):
    padLength += 8

  output = ""
  lineLength = 0
  for name in files:
    output += name.ljust(padLength)
    lineLength += padLength
    if (TERMINAL_LENGTH - lineLength < padLength):
      output += "\n"
      lineLength = 0
  
  return output

def printOutput(output):
  for string in output:
    if (string.startswith("ls.py:")):
      print(string, file = sys.stderr)
    else:
      print(string)

start()